base/shell.jl

```diff
@@ -253,3 +253,129 @@ julia> Base.shell_escape_posixly("echo", "this", "&&", "that")
+string for a Windows command line. Windows passes the entire command line as a single string to
+the application (unlike POSIX systems, where the list of arguments are passed separately).
+Many Windows API applications (including julia.exe), use the conventions of the [Microsoft C
+runtime](https://docs.microsoft.com/en-us/cpp/c-language/parsing-c-command-line-arguments) to
+split that command line into a list of strings. This function implements the inverse of such a
+C runtime command-line parser. It joins command-line arguments to be passed to a Windows console
+application into a command line, escaping or quoting meta characters such as space,
+double quotes and backslash where needed. This may be useful in concert with the `windows_verbatim`
+flag to [`Cmd`](@ref) when constructing process pipelines.
+
+# Example
+```jldoctest
+julia> println(shell_escaped_winsomely("A B\\", "C"))
+"A B\\" C
+"""
+shell_escape_winsomely(args::AbstractString...) =
+    sprint(print_shell_escaped_winsomely, args..., sizehint=(sum(length, args)) + 3*length(args))
+
+function print_shell_escaped_CMDly(io::IO, arg::AbstractString)
+    any(in("\r\n"), arg) && throw("Encountered unsupported character by CMD.")
+    # include " so to avoid toggling behavior of ^
+    # the rule for ! is a bit esoteric, but doesn't hurt to add it
+    arg = replace(arg, r"[%!^<>&|]" => s"^\0")
+    print(io, arg)
+end
+
+"""
+     shell_escape_CMDly(arg::AbstractString)::String
+
+The unexported `shell_escape_CMDly` function takes a string and escapes any special characters
+in such a way that it is safe to pass it as an argument to some `CMD.exe`. This may be useful
+in concert with the `windows_verbatim` flag to [`Cmd`](@ref) when constructing process
+pipelines.
+
+See also [`shell_escape_BATCHly`](@ref) and [`shell_escape_PWSHly`](@ref).
+
+# Example
+```jldoctest
+julia> println(shell_escape_CMDly("\"A B\\\" & C"))
+^"A B\\^" ^& C
+
+!important
+    Due to a peculiar behavior of the CMD, each command after a literal `|` character
+    (indicating a command pipeline) must have `shell_escape_CMDly` applied twice. For example:
+    ```
+    to_print = "All for 1 & 1 for all!"
+    run(Cmd(Cmd(["cmd /c \"break | echo \$(shell_escape_CMDly(shell_escape_CMDly(to_print)))"]), windows_verbatim=true))
+    ```
+"""
+shell_escape_CMDly(arg::AbstractString) = sprint(print_shell_escaped_CMDly, arg)
+
+function print_shell_escaped_BATCHly(io::IO, arg::AbstractString)
+    # see https://www.robvanderwoude.com/variableexpansion.php
+    # the rule for ! is a bit esoteric, but doesn't hurt to add it
+    any(in("\r\n"), arg) && throw("Encountered unsupported character by CMD")
+    arg = replace(arg, r"[()!^\"<>&|]" => s"^\0")
+    arg = replace(arg, "%" => "%%")
+    print(io, arg)
+end
+
+"""
+     shell_escape_BATCHly(arg::AbstractString)::String
+
+Like [`shell_escape_CMDly`](@ref), but appropriate for use inside batch (*.bat) files for use
+with `CMD.exe`.
```

Comment:
actually, this might be confusing to folks, so perhaps we should not have this function. I think I'd naively expect this to be the function to quote arguments to a batch script, while this actually quotes them to be values inside a batch script (and it seems much less likely that someone is going to emit `*.bat` files and then run them)